backend/postgres/postgres.go

```diff
@@ -0,0 +1,955 @@
+			maxDequeueCount = dequeueCount
+		}
+
+		e := new(protos.HistoryEvent)
+		err := protojson.Unmarshal(eventPayload, e)
+		if err != nil {
+			return nil, err
+		}
+
+		newEvents = append(newEvents, e)
+	}
+
+	if err = tx.Commit(); err != nil {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		return nil, fmt.Errorf("failed to update orchestration work-item: %w", err)
+	}
+
+	wi := &backend.OrchestrationWorkItem{
+		InstanceID: api.InstanceID(instanceID),
+		NewEvents:  newEvents,
+		LockedBy:   be.workerName,
+		RetryCount: maxDequeueCount - 1,
+	}
+
+	return wi, nil
+}
+
+func (be *postgresBackend) GetActivityWorkItem(ctx context.Context) (*backend.ActivityWorkItem, error) {
+	if err := be.ensureDB(ctx); err != nil {
+		return nil, err
+	}
+
+	now := time.Now().UTC()
+	newLockExpiration := now.Add(be.activityLockTimeout)
+
+	row := be.db.QueryRowContext(
+		ctx,
+		`UPDATE durabletask.NewTasks SET LockedBy = $1, LockExpiration = $2, DequeueCount = DequeueCount + 1
+		WHERE SequenceNumber = (
+			SELECT SequenceNumber FROM durabletask.NewTasks T
+			WHERE T.LockExpiration IS NULL OR T.LockExpiration < $3
+			ORDER BY SequenceNumber ASC
+			LIMIT 1
+		) RETURNING SequenceNumber, InstanceID, EventPayload`,
+		be.workerName,
+		newLockExpiration,
+		now,
+	)
+
+	if err := row.Err(); err != nil {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		return nil, fmt.Errorf("failed to query for activity work-items: %w", err)
+	}
+
+	var sequenceNumber int64
+	var instanceID string
+	var eventPayload []byte
+
+	if err := row.Scan(&sequenceNumber, &instanceID, &eventPayload); err != nil {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		if err == sql.ErrNoRows {
+			// No new activity tasks to process
+			return nil, backend.ErrNoWorkItems
+		}
+		return nil, fmt.Errorf("failed to scan the activity work-item: %w", err)
+	}
+
+	e := new(protos.HistoryEvent)
+	err := protojson.Unmarshal(eventPayload, e)
+	if err != nil {
+		return nil, err
+	}
+
+	wi := &backend.ActivityWorkItem{
+		SequenceNumber: sequenceNumber,
+		InstanceID:     api.InstanceID(instanceID),
+		NewEvent:       e,
+		LockedBy:       be.workerName,
+	}
+	return wi, nil
+}
+
+func (be *postgresBackend) CompleteActivityWorkItem(ctx context.Context, wi *backend.ActivityWorkItem) error {
+	if err := be.ensureDB(ctx); err != nil {
+		return err
+	}
+
+	tx, err := be.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	bytes, err := protojson.Marshal(wi.Result)
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.ExecContext(ctx, "INSERT INTO durabletask.NewEvents (InstanceID, EventPayload) VALUES ($1, $2)", string(wi.InstanceID), bytes)
+	if err != nil {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		return fmt.Errorf("failed to insert into NewEvents table: %w", err)
+	}
+
+	dbResult, err := tx.ExecContext(ctx, "DELETE FROM durabletask.NewTasks WHERE SequenceNumber = $1 AND LockedBy = $2", wi.SequenceNumber, wi.LockedBy)
```

Comment:
For funsies, you could perform both these queries in a single statement to avoid the use of a transaction, so you'd reduce the number of round-trips from 4 to 1 and avoid locking the tables... 

```sql
WITH del AS (
    DELETE FROM durabletask.NewTasks
        WHERE SequenceNumber = $1
        AND LockedBy = $2
)
INSERT INTO durabletask.NewEvents (InstanceID, EventPayload) VALUES ($3, $4)
```

See for example: https://dba.stackexchange.com/questions/267243/want-to-delete-then-insert-with-single-statement-using-a-cte-in-postgres